File: points_to_prints/python/edge_matching/criterion.py
from abc import ABC, abstractmethod
from typing import List

from fastquadtree import QuadTree

from .constants import *
from .geometry import Point, Segment


class Criterion(ABC):

    @abstractmethod
    def evaluate_segments(self, segments: List[Segment], perimeter: float) -> float:
        raise NotImplementedError("Subclasses must implement this method")


class LinearCriterion(Criterion):

    def __init__(
        self,
        points: list[Point],
        weights: list[float],
        max_distance: float,
        alpha_ratio: float,
        alpha_abs: float,
        initial_perimeter: float,
    ) -> None:
        # Check that weights and points have the same length
        if len(points) != len(weights):
            raise ValueError("Points and weights must have the same length")

        self.points = points
        self.weights = weights
        self.max_distance = max_distance
        self.alpha_ratio = alpha_ratio
        self.alpha_abs = alpha_abs

        min_x = min(point.x for point in points) - 1e-6
        max_x = max(point.x for point in points) + 1e-6
        min_y = min(point.y for point in points) - 1e-6
        max_y = max(point.y for point in points) + 1e-6
        self.qt = QuadTree((min_x, min_y, max_x, max_y), len(points))
        for i, point in enumerate(points):
            self.qt.insert((point.x, point.y), id_=i)

        self.initial_perimeter = initial_perimeter

    def evaluate_segments(self, segments: List[Segment], perimeter: float) -> float:
        proximity_value = 0.0
        for segment in segments:
            bounding_box = segment.bounding_box().buffer(self.max_distance)
            ids, coords = self.qt.query_np(
                (
                    bounding_box.min_x,
                    bounding_box.min_y,
                    bounding_box.max_x,
                    bounding_box.max_y,
                )
            )

            for point_idx in ids:
                point = self.points[point_idx]
                weight = self.weights[point_idx]
                normal_distance, segment_distance = segment.distance_to_point(point)
                if normal_distance > self.max_distance:
                    continue
                if segment_distance > 0:
                    continue
                proximity_value -= weight * (1.0 - normal_distance)

        if perimeter > self.initial_perimeter:
            perimeter_ratio_value = self.alpha_ratio * abs(
                perimeter / self.initial_perimeter
            )
        else:
            perimeter_ratio_value = self.alpha_abs * abs(
                self.initial_perimeter / perimeter
            )

        perimeter_abs_value = self.alpha_abs * perimeter

        return proximity_value + perimeter_ratio_value + perimeter_abs_value
        return proximity_value + perimeter_ratio_value + perimeter_abs_value
        return proximity_value + perimeter_ratio_value + perimeter_abs_value

File: points_to_prints/cpp/src/edge_matching/criterion.cpp
#include "criterion.hpp"

#include "../utils/cgal.hpp"
#include "constants.hpp"
#include <CGAL/enum.h>

double LinearCriterion::evaluate_segments(
    const std::vector<Segment_2> &segments,
    const std::vector<double> &segments_initial_length) const {
    // Check that the input vectors have the same size
    if (segments.size() != segments_initial_length.size()) {
        throw std::invalid_argument(
            "Segments and segments_initial_length must have the same size");
    }

    // Compute the perimeters
    double current_perimeter = 1e-6;
    double initial_perimeter = 1e-6;
    for (size_t i = 0; i < segments.size(); ++i) {
        current_perimeter += std::sqrt(segments.at(i).squared_length());
        initial_perimeter += segments_initial_length.at(i);
    }

    // Compute the proximity value
    std::vector<double> point_best_score(points.size(), 0.0);
    double proximity_value = 0.0;
    for (size_t i = 0; i < segments.size(); ++i) {
        const auto &segment = segments.at(i);

        UnitVector_2 segment_normal(
            segment.supporting_line().to_vector().perpendicular(
                CGAL::CLOCKWISE));

        // std::cout << "Evaluating segment " << i << ": " << segment <<
        // std::endl;

        // Query the points that may be close enough
        std::vector<std::size_t> nearby_point_indices =
            las_kd_tree.search_indices_in_box(segment.bbox(),
                                              EDGE_CRITERION_MAX_DISTANCE);

        // std::cout << "Number of nearby points: " <<
        // nearby_point_indices.size()
        //           << std::endl;

        // int num_points_not_on_segment = 0;
        // int num_points_far_from_segment = 0;
        // int num_points_considered = 0;

        for (std::size_t point_index : nearby_point_indices) {
            const Point_2 &point = points.at(point_index);
            double weight = weights.at(point_index);

            // Check if the projection of the point on the segment is within the
            // segment
            const Point_2 &closest_point =
                segment.supporting_line().projection(point);
            auto dist_proj_segment =
                CGAL::squared_distance(closest_point, segment);
            if (dist_proj_segment > 1e-6) {
                // num_points_not_on_segment++;
                continue;
            }

            // Check if the distance from the point to the segment smaller than
            // the threshold
            double distance = std::sqrt(CGAL::squared_distance(point, segment));
            if (distance > EDGE_CRITERION_MAX_DISTANCE) {
                // num_points_far_from_segment++;
                continue;
            }

            // Check the angle between the point normal and the segment normal
            UnitVector_2 point_normal = point_normals.at(point_index);
            double dot_product = point_normal * segment_normal;
            if (dot_product < 0.5) {
                continue;
            }
            weight *= dot_product;

            // Compute the proximity score
            double proximity_score =
                weight * (1.0 - distance / EDGE_CRITERION_MAX_DISTANCE);
            point_best_score[point_index] =
                std::max(point_best_score[point_index], proximity_score);

            // num_points_considered++;
        }
        // std::cout << "Not on segment: " << num_points_not_on_segment
        //           << ", Far from segment: " << num_points_far_from_segment
        //           << ", Considered: " << num_points_considered << std::endl;
    }
    proximity_value =
        -std::accumulate(point_best_score.begin(), point_best_score.end(), 0.0);
    proximity_value *= EDGE_CRITERION_POINT_DENSITY;

    // // Compute the perimeter ratio value
    // double perimeter_ratio_value;
    // if (current_perimeter > initial_perimeter) {
    //     perimeter_ratio_value =
    //         (current_perimeter / initial_perimeter) * initial_perimeter;
    // } else {
    //     perimeter_ratio_value =
    //         (initial_perimeter / current_perimeter) * initial_perimeter;
    // }
    // perimeter_ratio_value *= this->alpha_ratio;

    // Compute the mean edge length ratio value
    double mean_edge_length_ratio_value = 0.0;
    for (size_t i = 0; i < segments.size(); ++i) {
        double current_length = std::sqrt(segments.at(i).squared_length());
        double initial_length = segments_initial_length.at(i);
        if (current_length > initial_length) {
            mean_edge_length_ratio_value += (current_length / initial_length);
        } else {
            mean_edge_length_ratio_value += (initial_length / current_length);
        }
    }
    mean_edge_length_ratio_value /= segments.size();
    mean_edge_length_ratio_value *= this->alpha_ratio;

    // Compute the perimeter absolute value
    double perimeter_abs_value = this->alpha_absolute * current_perimeter;

    return proximity_value + mean_edge_length_ratio_value + perimeter_abs_value;
}